### DISCUSSION TITLE
How to read the `source-layer`s from a vector tile source (pbf)?

### DISCUSSION THREAD

**Discussion by tordans** [maplibre/maplibre-gl-js#2236] [Q&A]
Is there a way that Maplibre GL JS could read the available `source-layer` names from a vector tile source?

The source is required to style vector data, however it is more or less invisible and therefore really hard to get right. I am looking into way to make this part of the map making easier.

---

Related question: Would it be possible for Maplibre GL JS to have a "all sources" (AKA source is `undefined`) option, where given styles are applied to all available `source-layer`? 

I wonder if this would make a better default behavior since it would show styles more reliably and still provide a way to fix situations where multiple `source-layers` are present.

**Comment by HarelM** [comment]
Editors such as Maputnik show the source layer I believe, so styling a map using those tools would be the right way to go.
The second part is interesting, there's a different between a source and a source-layer, which one did you mean when you wrote "all sources"?
I'm not sure how performant it would be to apply a layer to all sources though, but an interesting idea non-the-less.

**Reply by tordans** [reply]
@HarelM 

> The second part is interesting, there's a different between a source and a source-layer, which one did you mean when you wrote "all sources"?

Most of my sources have just one source-layer. But the name is unknown or complicated to find out. So I was hoping for a way to tell the renderer to "just use the source-layer that are there and apply the styles", without explicitly specifying the source-layer name.

**Reply by HarelM** [reply]
Ahh, interesting, something like not specifying a source-layer in the style-layer for a vector source would do, like it does for geojson, and apply the style-layer on all the source layers, right?
I would recommend opening this modification request in the style-spec repo and hear what other people has to say about it.

**Comment by tordans** [comment]
I just learned about https://github.com/mapbox/vector-tile-js which seems to be a solution to read the layers or source-layers? from a given pbf file. This could solve my first issue, to show some helper debug info about how the source-layer are named.
